core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1755,6 +1755,22 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
       case None    => Pull.output1(fallback)
     }.stream
 
+  /** Emits the first `n` elements of this stream,
+    * raising an IllegalStateException if there are more elements.
+    */
+  def limit[F2[x] >: F[x]](n: Long)(implicit rt: RaiseThrowable[F2]): Stream[F2, O] =
+    this.pull
+      .take(n)
+      .flatMap {
+        case Some(s) =>
+          s.pull.uncons.flatMap {
+            case Some(_) => Pull.raiseError(new IllegalStateException)
```

Comment:
I suggest `new IllegalStateException(s"limit($n) emitted more than $n elements")` or something similar.

Reply:
Added in https://github.com/typelevel/fs2/pull/3113/commits/98141abd75393d9dc0ad60b9236770e34fd3a5ca